src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+const MAX_CACHE_ENTRIES: DefMinMax<u64> =
+    DefMinMax::new(1_000, 1, 1_000_000_000);
+
+/// Limit on the maximum time a cache entry is considered valid.
+///
+/// According to RFC 8767 the limit should be on the order of days to
+/// weeks with a recommended cap of 604800 seconds (7 days).
+const MAX_VALIDITY: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(604800),
+    Duration::from_secs(60),
+    Duration::from_secs(6048000),
+);
+
+/// Amount of time to cache transport failures.
+///
+/// According to RFC 9520 at least 1 second and at most 5 minutes.
+const TRANSPORT_FAILURE_DURATION: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(30),
+    Duration::from_secs(1),
+    Duration::from_secs(5 * 60),
+);
+
+/// Limit on the amount of time to cache DNS result codes that are not
+/// NOERROR or NXDOMAIN.
+///
+/// According to RFC 9520 at least 1 second and at most 5 minutes.
+const MISC_ERROR_DURATION: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(30),
+    Duration::from_secs(1),
+    Duration::from_secs(5 * 60),
+);
+
+/// Limit on the amount of time to cache a NXDOMAIN error.
+///
+/// According to RFC 2308 the limit should be one to three hour with a
+/// maximum of one day.
+const MAX_NXDOMAIN_VALIDITY: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(3600),
+    Duration::from_secs(60),
+    Duration::from_secs(24 * 3600),
+);
+
+/// Limit on the amount of time to cache a NODATA response.
+///
+/// According to RFC 2308 the limit should be one to three hour with a
+/// maximum of one day.
+const MAX_NODATA_VALIDITY: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(3600),
+    Duration::from_secs(60),
+    Duration::from_secs(24 * 3600),
+);
+
+/// Limit on the amount of time a delegation is considered valid.
+const MAX_DELEGATION_VALIDITY: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(1_000_000),
+    Duration::from_secs(60),
+    Duration::from_secs(1_000_000_000),
+);
+
+// The following four flags are relevant to caching: AD, CD, DO, and RD.
+//
+// The AD flag needs to be part of the key when DO is clear. When replying,
+// if both AD and DO are not set in the original request then AD needs to be
+// cleared if it was set in the response (extra look up if no entry with
+// AD clear exists.
+//
+// The CD flag partitions the cache, responses to request with CD set must not
+// be visible to requests with CD and vice versa.
+//
+// A request with DO set can only be satisfied with a response to a request
+// with CD set. However, if CD in the request is clear then a reponse to a
+// request with CD set can be used if all unrequested DNSSEC records are
+// stripped.
+//
+// A request with RD clear can be satisfied by a response to a request with
+// RD set. For simplicitly requests with RD set can only get a response to
+// a request with RD set. In theory some responses to requests with RD clear
```

Comment:
Again I find this use of request twice in the same sentence confusing, a response to a request to a response - eh?

Reply:
I hope the following is more clear:
// A request with RD clear can be satisfied by a response to a request with
// RD set. For simplicitly requests with RD set will only get a cached
// response to another request with RD set. In theory some responses to
// requests with RD clear could be used to satisfy requests with RD set.
// However, this is not implemented.